### PR TITLE
Unary prefix operators

### PR DESCRIPTION
**Pull request by Artem-Semenov-dev**
Now it's really done.

### CODE REVIEW

**Comment by Artem-Semenov-dev** [comment]
Merge reading and updating variables into one machine.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/ScriptElementExecutorFactoryImpl.java [hunk 1]
```diff
@@ -7,157 +7,98 @@
...
                 FiniteStateMachine.oneOfMachine(
-                        errorMessage -> {
-                            throw new ExecutionException(errorMessage);
-                        },
+                        exceptionThrower,
                         new ExecutorProgramElementTransducer(ScriptElement.INIT_VAR, this).named("Variable initialisation"),
                         new ExecutorProgramElementTransducer(ScriptElement.WHILE_OPERATOR, this).named("While loop"),
-                        new ExecutorProgramElementTransducer(ScriptElement.PROCEDURE, this).named("Procedure"))));
+                        new ExecutorProgramElementTransducer(ScriptElement.PROCEDURE, this).named("Procedure"),
+                        new ExecutorProgramElementTransducer(ScriptElement.PRODUCE_VARIABLE, this).named("UnaryPrefixOperator"))));
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/ScriptElementExecutorFactoryImpl.java]: What is PRODUCE_VARIABLE? Why is it named 'UnaryPrefixOperator'? Make it clear.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/executors/NumberExecutor.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+            if (output.isParseOnly()) {
+                return true;
+            }
+
+            output.systemStack().current().pushOperand(executeResult.get());
+
+            return true;
+        }
+
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/executors/NumberExecutor.java:28]: Make it simpler:

        if (executeResult.isPresent()) {

            if (!output.isParseOnly()) {
                output.systemStack().current().pushOperand(executeResult.get());;
            }

            return true;
        }

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/executors/ProcedureFactoryExecutor.java [hunk 3]
```diff
@@ -29,8 +29,11 @@ public boolean execute(CharSequenceReader inputChain, ScriptContext output) thro
 
         if (machine.run(inputChain, functionHolder)) {
 
-            procedureFactory.create(functionHolder.getFunctionName())
-                    .execute(functionHolder.getArguments(), functionHolder.getScriptContext());
+            if (procedureFactory.hasProcedure(functionHolder.getFunctionName())) {
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/executors/ProcedureFactoryExecutor.java:32]: FunctionHolder?)

Comment (Oleg-Melnik) [reply bazascript/src/main/java/com/teamdev/bazascript/interpreter/executors/ProcedureFactoryExecutor.java:32]: Why is this part of this PR?)

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/executors/RelationalExpressionElementExecutor.java [hunk 4]
```diff
@@ -39,12 +39,20 @@ public boolean execute(CharSequenceReader inputChain, ScriptContext output) thro
...
                                     }
                                 }).named("Binary operator"),
                         partOfExpression));
 
-        return relationalMachine.run(inputChain, output);
+        if (!relationalExpressionMachine.run(inputChain, output)) {
+
+            output.memory().clearCache();
+            output.memory().updateCache();
+            return false;
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/executors/RelationalExpressionElementExecutor.java:55]: What is this?)

            output.memory().clearCache();
            output.memory().updateCache();

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/executors/VariableExecutor.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import com.teamdev.fsm.FiniteStateMachine;
+import com.teamdev.fsm.Transducer;
+
+import java.util.List;
+
+/**
+ * {@code VariableExecutor} is an implementation of {@link ScriptElementExecutor}.
+ * {@code VariableExecutor} is used run produceVariableMachine for
+ * read variables or update them if they have unary prefix operators and pushed result to output.
+ */
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/executors/VariableExecutor.java:19]: Rewrite.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/program/ProgramMachine.java [hunk 6]
```diff
@@ -24,7 +24,14 @@ private ProgramMachine(TransitionMatrix<ProgramStates> matrix, ScriptElementExec
 
         registerTransducer(START, Transducer.illegalTransition());
         registerTransducer(STATEMENT, new StatementTransducer(factory.create(ScriptElement.STATEMENT)));
-        registerTransducer(SEPARATOR, Transducer.checkAndPassChar(';'));
+        registerTransducer(SEPARATOR, Transducer.<ScriptContext, ExecutionException>checkAndPassChar(';').and(
+                (inputChain, outputChain) -> {
+
+                    outputChain.memory().updateVariables();
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/program/ProgramMachine.java]: What is this? Provide a name with a sense.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/runtime/Memory.java [hunk 7]
```diff
@@ -1,37 +1,75 @@
...
+import com.teamdev.implementations.type.DoubleValueVisitor;
 import com.teamdev.implementations.type.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.HashMap;
 import java.util.Map;
 
 /**
  * {@code Memory} is a class that used to store Double type variables.
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/runtime/Memory.java]: Double only?)

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/runtime/Memory.java [hunk 8]
```diff
@@ -1,37 +1,75 @@
...
+ * Memory has cache that store variables in cases of finite state machine rollback.
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/runtime/Memory.java]: Why do you need a cache? Consider removal or provide very clear javadoc.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/util/FunctionTransducer.java [hunk 9]
```diff
@@ -10,8 +11,6 @@
 
     private final BiConsumer<O, Value> consumer;
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/util/FunctionTransducer.java]: 'consumer' of what?

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/util/UnaryPrefixOperatorContext.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+public class UnaryPrefixOperatorContext implements WithContext {
+
+    private final ScriptContext scriptContext;
+
+    private boolean readVariableOnly;
+
+    private UnaryOperator<Value> unaryOperator;
+
+    public UnaryPrefixOperatorContext(ScriptContext scriptContext) {
+        this.scriptContext = scriptContext;
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/util/UnaryPrefixOperatorContext.java]: Use Preconditions.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/util/UnaryPrefixOperatorContext.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * {@code UnaryPrefixOperatorContext} is an implementation of {@link WithContext}.
+ * {@code UnaryPrefixOperatorContext} can be used like an output chain for machine that execute expressions with unary prefix operator.
+ */
+
+public class UnaryPrefixOperatorContext implements WithContext {
+
+    private final ScriptContext scriptContext;
+
+    private boolean readVariableOnly;
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/util/UnaryPrefixOperatorContext.java]: What is this?

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/util/UnaryPrefixOperatorContext.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+    private UnaryOperator<Value> unaryOperator;
+
+    public UnaryPrefixOperatorContext(ScriptContext scriptContext) {
+        this.scriptContext = scriptContext;
+    }
+
+    public void setUnaryOperator(UnaryOperator<Value> unaryOperator) {
+        this.unaryOperator = unaryOperator;
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/util/UnaryPrefixOperatorContext.java]: Use Preconditions.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/util/UnaryPrefixOperatorContext.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        this.readVariableOnly = readVariableOnly;
+    }
+
+    public boolean readVariableOnly() {
+        return readVariableOnly;
+    }
+
+    public Value applyOperator(Value operand) {
+
+        return unaryOperator.apply(operand);
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/util/UnaryPrefixOperatorContext.java]: Use Preconditions.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/variable/ProduceVariableTransducer.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+                    outputChain.getScriptContext().systemStack().current().pushOperand(variableUpdateResult);
+                }
+
+                return true;
+
+            } else
+                throw new ExecutionException("Operation with not initialised variable " + variableName + " cannot be performed");
+        }
+        return false;
+    }
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/variable/ProduceVariableTransducer.java:56]: Try to make the logic of this method a bit clearer.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/variable/UnaryPrefixOperatorTransducer.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Override
+    public boolean doTransition(CharSequenceReader inputChain, UnaryPrefixOperatorContext outputChain) throws ExecutionException {
+        Preconditions.checkNotNull(inputChain, outputChain);
+
+        if (!inputChain.canRead()) {
+            return false;
+        }
+
+        Optional<UnaryOperator<Value>> unaryOperator = factory.create(inputChain.readOperator());
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/variable/UnaryPrefixOperatorTransducer.java]: InputChain can't have method 'readOperator()'.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/variable/UnaryPrefixOperatorTransducer.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+        if (unaryOperator.isPresent()) {
+
+            outputChain.setUnaryOperator(unaryOperator.get());
+
+            inputChain.incrementPosition();
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/variable/UnaryPrefixOperatorTransducer.java]: Why do you need it here? +1? Why? Logic is not clear.

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/whileoperator/WhileOperatorContext.java [hunk 17]
```diff
@@ -11,24 +11,24 @@ public class WhileOperatorContext implements WithContext {
 
     private int position;
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/whileoperator/WhileOperatorContext.java]: Position of what?

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/whileoperator/WhileOperatorContext.java [hunk 18]
```diff
@@ -11,24 +11,24 @@ public class WhileOperatorContext implements WithContext {
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/whileoperator/WhileOperatorContext.java:11]: Condition value?

File: bazascript/src/main/java/com/teamdev/bazascript/interpreter/whileoperator/WhileOperatorContext.java [hunk 19]
```diff
@@ -11,24 +11,24 @@ public class WhileOperatorContext implements WithContext {
...
 
-    public void setCondition(boolean condition) {
-        this.condition = condition;
+    int getPosition() {
+        return position;
     }
 
-    public void setPosition(int position) {
+    void setPosition(int position) {
         this.position = position;
```
Comment (Oleg-Melnik) [inline bazascript/src/main/java/com/teamdev/bazascript/interpreter/whileoperator/WhileOperatorContext.java]: Use Preconditions.

File: bazascript/src/test/java/com/teamdev/bazascript/interpreter/InterpreterTest.java [hunk 20]
```diff
@@ -7,19 +7,20 @@
 import static org.junit.jupiter.params.provider.Arguments.of;
 
 class InterpreterTest extends AbstractTest {
```
Comment (Oleg-Melnik) [inline bazascript/src/test/java/com/teamdev/bazascript/interpreter/InterpreterTest.java:9]: Separate the test by testing elements.

File: bazascript/src/test/java/com/teamdev/bazascript/interpreter/UnaryPrefixOperatorTest.java [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+
+                of("++a;", 3, "Unary prefix operator with uninitialised variable test has not throw exception"),
+
+                of("+a;", 0, "Unary prefix operator with only one plus sign test has not throw exception"),
+
+                of("++(5+2);", 0, "Unary prefix operator with expression test has not throw exception"),
+
+                of("a = 0; ++(a); print(a)", 7, "Unary prefix operator with variable in brackets test has not throw exception")
+
+        );
```
Comment (Oleg-Melnik) [inline bazascript/src/test/java/com/teamdev/bazascript/interpreter/UnaryPrefixOperatorTest.java:50]: Add more tests and use DOT at the end of sentences.

File: fsm/src/main/java/com/teamdev/fsm/CharSequenceReader.java [hunk 22]
```diff
@@ -29,25 +30,26 @@ public String readOperator(){
```
Comment (Oleg-Melnik) [inline fsm/src/main/java/com/teamdev/fsm/CharSequenceReader.java]: This method can't be here. CharSequenceReader knows nothing about operators.

File: fsm/src/main/java/com/teamdev/fsm/CharSequenceReader.java [hunk 23]
```diff
@@ -29,25 +30,26 @@ public String readOperator(){
...
-        else operator.append(read());
+
+        if (startPosition != position())
+            decrementPosition();
 
         return operator.toString();
     }
 
+    public boolean isOperator(char sign){
+
```
Comment (Oleg-Melnik) [inline fsm/src/main/java/com/teamdev/fsm/CharSequenceReader.java:36]: This method can't be here. CharSequenceReader knows nothing about operators.

File: fsm/src/main/java/com/teamdev/fsm/CharSequenceReader.java [hunk 24]
```diff
@@ -29,25 +30,26 @@ public String readOperator(){
...
+        List<Character> operators = List.of('+', '-', '>', '<', '=', '%', '^', '*', '/');
```
Comment (Oleg-Melnik) [inline fsm/src/main/java/com/teamdev/fsm/CharSequenceReader.java]: Only factory may know the signs of operators.

File: implementations/src/main/java/com/teamdev/implementations/machines/expression/ExpressionMachine.java [hunk 25]
```diff
@@ -25,6 +25,7 @@ public static <O, E extends Exception> ExpressionMachine<O, E> create(BiConsumer
         TransitionMatrix<ExpressionStates> matrix = TransitionMatrix.<ExpressionStates>builder()
                 .withStartState(ExpressionStates.START)
                 .withFinishState(ExpressionStates.FINISH)
+                .withTemporaryState(ExpressionStates.OPERAND)
```
Comment (Oleg-Melnik) [inline implementations/src/main/java/com/teamdev/implementations/machines/expression/ExpressionMachine.java]: Why is OPERAND temporary? This breaks my brain.

File: implementations/src/main/java/com/teamdev/implementations/operators/PrefixUnaryOperatorFactory.java [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+
+        prefixUnaryOperators.put("--", value -> {
+
+            Double operand = DoubleValueVisitor.read(value);
+
+            --operand;
+
+            return new DoubleValue(operand);
+        });
+    }
```
Comment (Oleg-Melnik) [inline implementations/src/main/java/com/teamdev/implementations/operators/PrefixUnaryOperatorFactory.java:26]: 2 of 3 lines of code are the same... Remove code duplication.

**Review by Oleg-Melnik — CHANGES_REQUESTED** [review]

**Pull request merged**